src/eredis_cluster_manager.erl

```diff
@@ -0,0 +1,415 @@
+    ConnectedSlotsMaps = connect_all_slots(SlotsMaps),
+    Slots = create_slots_cache(ConnectedSlotsMaps),
+
+    NewState = State#state{
+        slots = list_to_tuple(Slots),
+        slots_maps = list_to_tuple(ConnectedSlotsMaps),
+        version = State#state.version + 1
+    },
+    NewState.
+
+-spec get_cluster_slots([#node{}]) -> [[bitstring() | [bitstring()]]].
+get_cluster_slots([]) ->
+    throw({error,cannot_connect_to_cluster});
+get_cluster_slots([Node|T]) ->
+    case lookup_eredis_pid(Node#node.address, Node#node.port) of
+        {ok, Pid} ->
+          case eredis:q(Pid, ["CLUSTER", "SLOTS"]) of
+            {error,<<"ERR unknown command 'CLUSTER'">>} ->
+                get_cluster_slots_from_single_node(Node);
+            {error,<<"ERR This instance has cluster support disabled">>} ->
+                get_cluster_slots_from_single_node(Node);
+            {ok, ClusterInfo} ->
+                ClusterInfo;
+            _ ->
+                get_cluster_slots(T)
+          end;
+        _ ->
+            get_cluster_slots(T)
+  end.
+
+
+-spec get_cluster_slots_from_single_node(#node{}) -> [[bitstring() | [bitstring()]]].
+get_cluster_slots_from_single_node(Node) ->
+    [[<<"0">>, integer_to_binary(?REDIS_CLUSTER_HASH_SLOTS-1),
+    [list_to_binary(Node#node.address), integer_to_binary(Node#node.port)]]].
+
+
+-spec parse_cluster_slots([[bitstring() | [bitstring()]]]) -> [#slots_map{}].
+parse_cluster_slots(ClusterInfo) ->
+    parse_cluster_slots(ClusterInfo, 1, []).
+
+parse_cluster_slots([[StartSlot, EndSlot | [[Address, Port | _] | _]] | T], Index, Acc) ->
+    SlotsMap =
+        #slots_map{
+            index = Index,
+            start_slot = binary_to_integer(StartSlot),
+            end_slot = binary_to_integer(EndSlot),
+            node = #node{
+                address = binary_to_list(Address),
+                port = binary_to_integer(Port)
+            }
+        },
+    parse_cluster_slots(T, Index+1, [SlotsMap | Acc]);
+parse_cluster_slots([], _Index, Acc) ->
+    lists:reverse(Acc).
+
+
+%% Creates tuple with one element per Redis slot. Each element maps the Redis slot to the Redis
+%% node index in SlotsMaps.
+-spec create_slots_cache([#slots_map{}]) -> [integer()].
+create_slots_cache(SlotsMaps) ->
+  SlotsCache = [[{Index, SlotsMap#slots_map.index}
+        || Index <- lists:seq(SlotsMap#slots_map.start_slot, SlotsMap#slots_map.end_slot)]
+        || SlotsMap <- SlotsMaps],
+  SlotsCacheF = lists:flatten(SlotsCache),
+  SortedSlotsCache = lists:sort(SlotsCacheF),
+  [ Index || {_,Index} <- SortedSlotsCache].
+
+
+-spec connect_all_slots([#slots_map{}]) -> [integer()].
+connect_all_slots(SlotsMapList) ->
+    [SlotsMap#slots_map{node=connect_node(SlotsMap#slots_map.node)} || SlotsMap <- SlotsMapList].
+
+
+-spec connect_node(#node{}) -> #node{} | undefined.
+connect_node(Node) ->
+    case lookup_eredis_pid(Node#node.address, Node#node.port) of
+        {ok, _} -> Node;
+        _ -> undefined
+    end.
```

Comment:
why are we returning `undefined` here if we are unable to connect? shouldn't we log an error?